### PR TITLE
AIRNEIS-31: Setup backend tests

### PR DESCRIPTION
**Pull request by cedricgautier**
### Tests in Go

To be able to have tests in our codebase, we'll be using Gingko and Gomega which are popular test packages for Go. 

To start a test, you need to go inside the current folder on which you'll be testing your code and run the command: `go test`
When executed you will see if your test has passed or not and at the same time an XML file will be generated of the test file. 
---
### When your tests passes, it should look like this : 
Screenshot 2024-01-16 at 18 59 02

--- 
### When your tests fail, it should look like this : 
Screenshot 2024-01-16 at 19 00 37

### CODE REVIEW

File: back/lib/shared/hello/hello_test.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	. "airneis/lib/shared/hello"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/reporters"
+	. "github.com/onsi/gomega"
+)
+
+func TestHello(t *testing.T) {
+	RegisterFailHandler(Fail)
```
Comment (barnabepilliaudin) [inline back/lib/shared/hello/hello_test.go:13]: Can you please explain this line ?

File: back/lib/shared/hello/hello.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+	"strings"
+)
+
+func HelloWorld() (string, int) {
+	return "HelloWorld", 25
+}
+
+func returnLastname(fullName string) string {
+	return strings.Split(fullName, " ")[1]
+}
```
Comment (barnabepilliaudin) [inline back/lib/shared/hello/hello.go]: Could you remove this file in another commit as it's content is on the readme
But thx for providing test sample

File: back/lib/shared/hello/hello_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+var _ = Describe("HelloTest", func() {
+	Context("When we want to print hello", func() {
+		word, number := HelloWorld()
+		It("should return a string with hello world and 25 as an integer", func() {
+			Expect(word).To(Equal("HelloWorld"))
+			Expect(number).To(Equal(13))
+		})
+	})
+})
```
Comment (barnabepilliaudin) [inline back/lib/shared/hello/hello_test.go:25]: Could you remove this file in another commit as it's content is on the readme
But thx for providing test sample

File: docs/back/tests-in-golang.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+			Expect(number).To(Equal(13))
+		})
+	})
+})
+```
+
+### How to run tests
+
+- To run all tests in your current directory: `go test`
+- To run all tests in your current directory and sub-directories: `go test -v ./...`
```
Comment (barnabepilliaudin) [inline docs/back/tests-in-golang.md:126]: If in the root of backend directory will go test run all the tests or is there a command to run all the tests.
If yes can you put it in this file please

**Review by BarnabePILLIAUDIN — CHANGES_REQUESTED** [review]
- Is there a way to remove this warning ?
image
- Also the test report has been generated inside of the folder i ran the tests into is the a possiblity to have them in a dedicate file like the one in the gitIgnore

Except that it worked pretty well

File: back/lib/shared/hello/hello_test.go (hunk 1 above)
Comment (cedricgautier) [reply back/lib/shared/hello/hello_test.go:13]: The RegisterFailHandler connects Ginkgo to Gomega. When a matcher fails the fail handler passed into RegisterFailHandler is called.

File: back/lib/shared/hello/hello_test.go (hunk 3 above)
Comment (cedricgautier) [reply back/lib/shared/hello/hello_test.go:25]: Sure I can!

Comment (cedricgautier) [reply back/lib/shared/hello/hello_test.go:25]: I think to I'll keep it for now to not remove the packages when making a "go mod tidy", I'll remove it in the next PR when I add other real tests cases.

File: back/lib/shared/hello/hello.go (hunk 2 above)
Comment (cedricgautier) [reply back/lib/shared/hello/hello.go]: Same as below, I'll remove it in the next PR when I add other real tests cases.

File: docs/back/tests-in-golang.md (hunk 4 above)
Comment (cedricgautier) [reply docs/back/tests-in-golang.md:126]: You can use `go test -v ./...` from the root backend directory to run all tests

File: back/lib/shared/hello/hello_test.go (hunk 1 above)
Comment (barnabepilliaudin) [reply back/lib/shared/hello/hello_test.go:13]: Ok

File: back/lib/shared/hello/hello.go (hunk 2 above)
Comment (barnabepilliaudin) [reply back/lib/shared/hello/hello.go]: Ok

File: docs/back/tests-in-golang.md (hunk 4 above)
Comment (barnabepilliaudin) [reply docs/back/tests-in-golang.md:126]: Ok mb I misunderstood the last line

**Review by BarnabePILLIAUDIN — DISMISSED** [review]
LGTM

**Review by Magali-D — APPROVED** [review]

**Review by BarnabePILLIAUDIN — APPROVED** [review]
LGTM!

File: back/lib/shared/hello/hello.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package hello
+
+func HelloWorld() (string, int) {
+	return "HelloWorld", 25
+}
```
Comment (avetisk) [inline back/lib/shared/hello/hello.go:5]: If this is a template test or an example, it should be moved elsewhere, if it's not to be kept, then why has it been tracked?

File: back/go.mod [hunk 6]
```diff
@@ -4,35 +4,34 @@ go 1.21.3
...
 	github.com/valyala/fasthttp v1.50.0 // indirect
 	github.com/valyala/tcplisten v1.0.0 // indirect
+	golang.org/x/net v0.19.0 // indirect
 	golang.org/x/sys v0.15.0 // indirect
 	golang.org/x/text v0.14.0 // indirect
+	golang.org/x/tools v0.16.0 // indirect
 	google.golang.org/protobuf v1.32.0 // indirect
+	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
+	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
```
Comment (avetisk) [inline back/go.mod:36]: `go mod tidy` during CI?

File: back/lib/shared/hello/hello.go (hunk 5 above)
Comment (barnabepilliaudin) [reply back/lib/shared/hello/hello.go:5]: We have updated coding-practices.md in an other PR to avoid this situation in the future.

**Pull request merged**